### PR TITLE
feat: add KdV PINN sample crate (kdv_pinn)

### PR DESCRIPTION
**Pull request by terasakisatoshi**
## Summary

This PR adds a new sample crate `kdv_pinn` that demonstrates how to build a Physics-Informed Neural Network (PINN) for the Korteweg–de Vries (KdV) equation using tenferro-rs `TracedTensor` graphs.

- `sampler.rs`: generates collocation, initial-condition, and boundary-point tensors.
- `network.rs`: placeholder-based `Linear` / `Mlp` layers that are compiled once and executed many times.
- `pde.rs`: KdV residual `u_t + u*u_x + u_xxx` computed via repeated JVPs (since `TracedTensorAdExt::grad` requires a scalar output).
- `loss.rs`: MSE helpers and composite PINN loss.
- `optimizer.rs`: in-place SGD stepper for `Tensor` buffers.
- `main.rs`: builds the graph, compiles loss + per-parameter gradient programs, runs the training loop, and evaluates L2 relative error at `t = 0.5`.

## Key design notes / deviations from plan

- `TracedTensorAdExt::grad` only supports scalar outputs, so per-element derivatives are obtained through `jvp` with tangent tensors of ones.
- Learning rate was lowered from `0.01` to `0.001` because the full PDE + boundary objective diverges to NaN at the higher rate.
- `Sampler::collocation` returns `(Tensor, Tensor)` directly to avoid fragile buffer slicing in the training loop.

## Test plan

- `cargo fmt --all --check` ✅
- `cargo clippy -p kdv_pinn -- -D warnings` ✅
- `cargo test -p kdv_pinn` ✅ 12/12 tests pass
- `cargo test --workspace --release` ✅ passes
- `cargo run -p kdv_pinn --release` ✅ completes 500 epochs and prints final loss + L2 error

## Residual risk

The L2 relative error at `t = 0.5` after 500 epochs is around `1.0` (≈100%). The loss decreases from O(100) to O(1), gradients flow, and all unit tests pass, but the network has not fully converged to the analytical soliton. Further hyperparameter/architecture tuning is left to follow-up work.

## Work log

See `docs/worklogs/2026-06-15-kdv-pinn-sample.md` for the full session context.

### CODE REVIEW

**Comment by terasakisatoshi** [comment]
Update after tuning:

- Fixed the KdV nonlinearity coefficient: the reference soliton `u = 2 sech^2(x - 4t)` satisfies the standard equation `u_t + 6 u u_x + u_xxx = 0`, not `u_t + u u_x + u_xxx = 0`. This was the main convergence blocker.
- Switched the full training loop from SGD to Adam and added an `Adam` optimizer to `optimizer.rs`.
- Tuned hyperparameters: MLP `[2, 64, 64, 1]`, `N_COL = 512`, balanced loss weights, 1000 epochs, `LR = 1e-3`.
- Added regression tests for JVP-based derivatives and a residual check against the exact soliton.

Latest run:
- `cargo test -p kdv_pinn` ✅ 21/21 tests pass
- `cargo run -p kdv_pinn --release` ✅ final loss ~5e-3, L2 relative error at `t = 0.5` ~0.15 (down from ~1.0).

Full workspace tests are running.

**Comment by terasakisatoshi** [comment]
`cargo test --workspace --release` ✅ passes.

**Comment by terasakisatoshi** [comment]
Re-enabled auto-merge after the latest push (force-push had temporarily closed the PR).

The generated `kdv_pinn.gif` is excluded from the PR by `kdv_pinn/.gitignore`; it can be produced locally with:

```bash
cargo run -p kdv_pinn --release -- --gif kdv_pinn.gif
```

**Comment by terasakisatoshi** [comment]
Switched the GIF rendering from a hand-rolled Bresenham line drawer to `plotters`. The result now has proper axes, grid, legend, caption, and anti-aliased lines. Generated locally with `cargo run -p kdv_pinn --release -- --gif kdv_pinn.gif`.
